README
NAME
    Printer::Thermal - Interface for Thermal and some dot-matrix and inkjet
    Printers that support ESC-POS specification.

VERSION
    version 0.27

SYNOPSIS
       use Printer::Thermal;
 
       #For Network Printers $port is 9100 in most cases but might differ depending on how you have configured your printer
       $printer = Printer::Thermal->new(device_ip=>$printer_ip,device_port=>$port);
 
       #These commands won't actually send anything to the printer but it will store all the merged data including control codes to send to printer in $printer->print_string variable.
       $printer->write("Blah Blah \nReceipt Details\nFooter");
       $printer->bold_on();
       $printer->write("Bold Text");
       $printer->bold_off();
       $printer->print(); ##Sends the above set of code to the printer. Clears the buffer text in module.
 
       #For local printer connected on serial port, check syslog(Usually under /var/log/syslog) for what device file was created for your printer when you connect it to your system(For plug and play printers).
       my $path = '/dev/ttyACM0';
       $printer = Printer::Thermal->new(serial_device_path=$path);
       $printer->write("Blah Blah \nReceipt Details\nFooter");
       $printer->bold_on();
       $printer->write("Bold Text");
       $printer->bold_off();
       $printer->print();
 
       #For local printer connected on usb port, check syslog(Usually under /var/log/syslog) for what device file was created for your printer when you connect it to your system(For plug and play printers).
       my $path = '/dev/usb/lp0';
       $printer = Printer::Thermal->new(usb_device_path=$path);
       $printer->write("Blah Blah \nReceipt Details\nFooter");
       $printer->bold_on();
       $printer->write("Bold Text");
       $printer->bold_off();
       $printer->print();

DESCRIPTION
    Some might not find the module name accurate since ESC-P was developed
    initially for dot matrix and inkjet printers, however today most Thermal
    Receipt Printers use these codes for control. Most people(i.e. like me
    when I started looking for Thermal Printer stuff) who look for Thermal
    Printer codes don't know Thermal Printers use certain set of ESC codes
    to achieve a bunch of functions, and I didn't want to name it
    Printer::ESC::P because that would not help people who are new to
    receipt printers looking for something like this module. This module
    provides an Object oriented interface for interacting with Thermal
    Printers. Maybe I will refactor it later with subclasses. I used Moose
    and I apologize for that!!

    For ESC-P codes refer the guide from Epson
    http://support.epson.ru/upload/library_file/14/esc-p.pdf

ATTRIBUTES
  usb_device_path
    This variable contains the path for the printer device file when
    connected as a usb device on UNIX-like systems. I haven't added support
    for Windows and it probably wont work in doz as a local printer without
    some modifications. Feel free to try it out and let me know what
    happens. This must be passed in the constructor

  serial_device_path
    This variable contains the path for the printer device file when
    connected as a serial device on UNIX-like systems. I haven't added
    support for Windows and it probably wont work in doz as a local printer
    without some modifications. Feel free to try it out and let me know what
    happens. This must be passed in the constructor

  device_ip
    Contains the IP address of the device when its a network printer. The
    module creates IO:Socket::INET object to connect to the printer. This
    can be passed in the constructor.

  device_port
    Contains the network port of the device when its a network printer. The
    module creates IO:Socket::INET object to connect to the printer. This
    can be passed in the constructor.

  baudrate
    When used as a local serial device you can set the baudrate of the
    printer too. Default (38400) will usually work, but not always.

    This param may be specified when creating printer object to make sure it
    works properly.

    $printer = Printer::Thermal->new(serial_device_path => '/dev/ttyACM0',
    baudrate => 9600);

  read_char_time
    *DECRECATED*

  read_const_time
    *DECRECATED*

  black_threshold
    Black ink threshold, This param may be specified when creating the
    printer object. Default is 48.

  alpha_threshold
    Alpha threshold, This param may be specified when creating the printer
    object. Default is 127.

  heatTime
    Heating time to set for Supported Thermal Printers, this affects dot
    intensity.

    This param may be specified when creating the printer object. Default is
    120

  heatInterval
    This param may be specified when creating the printer object. Default is
    50

  heatingDots
    This param may be specified when creating the printer object. Default is
    7

  printer
    This is the direct device handle to the printer, You must almost never
    use this. Unless you are hacking through the module. If you are using
    this you must send me a bug report on why you had to use this.

    You can access it with $printer->printer

  print_string
    This contains the string in the module buffer that will be sent to the
    printer when you call $printer->print();

    my $print_string = $printer->print_string

  font
    Set ESC-POS Font

  underline
    Set/unset underline property

  emphasized
    Set/unset emphasized property

  double_height
    set unset double height property

  double_width
    set unset double width property

METHODS
  print
    $printer->print() Sends the accumulated commands to the printer. All
    commands below need to be followed by a print() to send the data from
    buffer to the printer. You may call more than one printer command and
    then call print to send them all to printer together. The following
    bunch of commands print a text to a printer, move down one line, and cut
    the receipt paper.

        $printer->write("hello Printer\n");
        $printer->linefeed();
        $printer->cutpaper();
        $printer->print(); # Sends the all the commands before this to the printer in one go.

  write
    $printer->write("some text\n") Writes a bunch of text that you pass here
    to the module buffer. Note that this will not be passed to the printer
    till you call $printer->print()

  left_margin
    $printer->left_margin($nl,$nh) Sets the left margin code to the printer.
    takes two single byte parameters, $nl and $nh. To determine the value of
    these two bytes, use the INT and MOD conventions. INT indicates the
    integer (or whole number) part of a number, while MOD indicates the
    remainder of a division operation. For example, to break the value 520
    into two bytes, use the following two equations: nH = INT 520/256 nL =
    MOD 520/256

  reset
    Resets the printer

  right_side_character_spacing
    Takes a one byte number, spacing as a parameter

  horiz_tab
    Adds a horizontal tab character like a \t to the print string.

  line_spacing
    Allows you to set the line spacing for the printer.

  linefeed
    Sends a new line character, i.e carriage return and line feed

  justify
    $alignment can be either 'L','C' or 'R' for left center and right
    justified printing

  bold_off
    Turns bold printing off

  bold_on
    Turns bold printing on

  doublestrike_off
    Turns doublestrike on characters off

  doublestrike_on
    Turns doublestrike on characters on

  emphasize_off
    Turns off emphasize(read ESC-POS documentation)

  emphasize_on
    Turns on emphasize(read ESC-POS documentation)

  font_size
    Defined Region 0 <= n <= 255 However, 1 <= vertical direction
    magnification ratio <= 8, 1 <= horizontal direction magnification ratio
    <= 8 Initial Value n=0 Function Specifies the character size
    (magnification ratio in the vertical and horizontal directions).

  font_size_esc
    Set ESC specified font size

        $printer->font_size_esc($size);

  font_b
    Switches printing to font b

  font_a
    Switches printing to font a

  underline_off
    Switches off underline

  underline_on
    Switches on underline

  inverse_off
    Switches off inverse text

  inverse_on
    Switches on inverse text

  barcode_height
    Sets barcode height

  print_barcode
    $printer->print_barcode($type,$string) Prints barcode

  print_text
    $printer->print_text($msg,$chars_per_line); Prints some text defined by
    msg. If chars_per_line is defined, inserts newlines after the given
    amount. Use normal '\n' line breaks for empty lines.

  print_bitmap
    To be done: This function is not implemented yet.

  color_1
    Prints in first color for dual color printers

  color_2
    Prints in second color for dual color printers

  cutpaper
    Cuts the paper. Most Thermal receipt printers support the facility to
    cut the receipt using this command once printing is done.

  open_cash_drawer
    Opens the Cash Drawer connected to the thermal printer.

  test
    Prints a bunch of test strings to see if your printer is working
    fine/connected properly. Don't worry if some things like emphasized and
    double strike looks the same, it happened with my printer too.

NOTES
    *   If the printer prints out garbled characters instead of proper text,
        try specifying the baudrate parameter when creating printer object
        when you create the printer object(not for network or USB printers)

         $printer = Printer::Thermal->new(serial_device_path => '/dev/ttyACM0', baudrate => 9600);

USAGE
    *   This Module offers a object oriented interface to ESC-POS Printers.

    *   Create a printer object by providing parameters for one of the three
        types of printers supported.

    *   then call formatting options or write() text to printer object in
        sequence.

    *   Then call the print() method to dispatch the sequences from the
        module buffer to the printer.

    Note: While you may call print() after every single command code, this
    is not advisable as some printers tend to choke up if you send them too
    many commands too quickly.

SUPPORT
  Bugs / Feature Requests
    Please report any bugs or feature requests through github at
    <https://github.com/shantanubhadoria/printer-thermal/issues>. You will
    be notified automatically of any progress on your issue.

  Source Code
    This is open source software. The code repository is available for
    public review and contribution under the terms of the license.

    <https://github.com/shantanubhadoria/printer-thermal>

      git clone git://github.com/shantanubhadoria/printer-thermal.git

AUTHOR
    Shantanu Bhadoria (shantanu@cpan.org)

CONTRIBUTOR
    Shantanu Bhadoria <shantanu att cpan dott org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Shantanu Bhadoria.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Changelog for Printer-Thermal

0.27 2013-11-25T03:46:48
 - Reduced Printer buffer size to accomadate printers with smaller buffers.
   [Shantanu]

0.26 2013-10-16T05:43:30
 - v0.26
   
    - minor edits [Shantanu] [Shantanu]
 - minor edits [Shantanu]

0.25 2013-10-16T04:51:44
 - v0.25 [Shantanu Bhadoria]

0.24 2013-10-16T04:27:09
 - v0.24 [Shantanu Bhadoria]

0.23 2013-10-16T04:23:10
 - v0.23 [Shantanu Bhadoria]
 - Fixed pod nested encoding error [Shantanu Bhadoria]

0.22 2013-10-14T06:59:45
 - v0.22 [Shantanu Bhadoria]
 - Testing wierdness in POD checker behaviour wrt podlint when testing >
   and < signs podlint expects them to be escaped but perldoc doesnt like
   it [Shantanu Bhadoria]

0.21 2013-10-14T06:16:42
 - v0.21 [Shantanu Bhadoria]
 - Fixed More POD Errors [Shantanu Bhadoria]

0.20 2013-09-30T07:58:17
 - v0.20 [Shantanu Bhadoria]
 - Attempting to fix Pod [Shantanu Bhadoria]
 - Adding mailmap to avoid explosions in contributors [Shantanu Bhadoria]

0.19 2013-09-15T05:34:37
 - v0.19
   
    - Fixed POD Errors [Shantanu] [Shantanu]
 - Fixed POD Errors [Shantanu]

0.18 2013-09-14T04:29:29
 - v0.18 [Shantanu]

0.17 2013-09-14T03:44:55
 - v0.17
   
    - Renamed example to examples to match spec [Shantanu] [Shantanu]
 - Renamed example to examples to match spec [Shantanu]

0.16 2013-09-14T01:50:04
 - v0.16
   
    - Fixed Pos Errors [Shantanu] [Shantanu]
 - Fixed Pos Errors [Shantanu]

0.15 2013-09-12T06:52:21
 - v0.15
   
    - Bump Changes [Shantanu Bhadoria] [Shantanu]
 - Bump Changes [Shantanu Bhadoria]

0.14 2013-07-11T03:40:42
 - v0.14
   
	   - 2013-07-11 11:40:17 Asia/Singapore
   
     - Code Cleanup added YML to ignore list [Shantanu Bhadoria]
 - Code Cleanup added YML to ignore list [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.13 2013-07-03T08:35:18
 - v0.13
   
	   - 2013-07-03 16:34:47 Asia/Singapore [Shantanu Bhadoria]
 - Fixed Time Hires include bug [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.12 2013-07-03T06:00:25
 - v0.12
   
	   - 2013-07-03 14:00:00 Asia/Singapore [Shantanu Bhadoria]
 - Added stopwords to avoid failing pod spell test [Shantanu Bhadoria]
 - Fixed POD spellings [Shantanu Bhadoria]
 - Added Buffering for Print output to avoid buffer overflow issues for
   long bills [Shantanu Bhadoria]
 - Fixed Documentation a little. Added a few more descriptions, fixed some
   stupid grammer [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.11 2013-06-14T05:08:10
 - v0.11
   
	   - 2013-06-14 13:07:46 Asia/Singapore [Shantanu Bhadoria]
 - Fixed Issues with overlapping bold and underline patterns [Shantanu
   Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.10 2013-06-11T02:02:20
 - v0.10
   
	   - 2013-06-11 09:34:09 Asia/Singapore [Shantanu Bhadoria]
 - Fixed POD Error encoding utf-8 [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.09 2013-06-11T01:28:12
 - v0.09
   
	   - 2013-06-11 09:27:41 Asia/Singapore [Shantanu Bhadoria]
 - Updated documentation and added notes and warnings (from: Peet de Vos)
   [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.08 2013-05-31T04:42:48
 - v0.08
   
	   - 2013-05-31 12:42:18 Asia/Singapore [Shantanu Bhadoria]
 - Added Drawer kick function [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]

0.07 2013-05-23T06:25:11
 - v0.07
   
	   - 2013-05-23 14:24:25 Asia/Singapore [Shantanu Bhadoria]

0.06 2013-05-23T05:08:22
 - Bump Changes [Shantanu Bhadoria]
 - vv0.5
   
	   - 2013-05-23 13:07:04 Asia/Singapore [Shantanu Bhadoria]
 - Dependency Hell [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]
 - vv0.4
   
	   - 2013-03-05 22:06:58 Asia/Singapore [Shantanu Bhadoria]
 - MOdified documentation for Pod::Weaver [Shantanu Bhadoria]
 - Bump Changes [Shantanu Bhadoria]
 - vv0.3
   
	   - 2013-03-05 21:58:48 Asia/Singapore [Shantanu Bhadoria]
 - Adde README.pod [Shantanu Bhadoria]
 - New files [Shantanu Bhadoria]
 - Deleted Unnecessary Files [Shantanu Bhadoria]
 - Adde Repo [Shantanu Bhadoria]
 - Updated README [Shantanu Bhadoria]
 - Removed BBCode prereq and refactored Documentation [Shantanu Bhadoria]
 - Modifications to library behaviour for multiple edits [Shantanu
   Bhadoria]
 - Fixed Example [Shantanu Bhadoria]
 - Updated Makefile with new dependencies [Shantanu Bhadoria]
 - Adding example script and segregating Serial and USB printers into
   seperate categories [Shantanu Bhadoria]
 - Add README files [Shantanu Bhadoria]
 - v0.01
   
   Added Todo Details [Shantanu Bhadoria]
 - First Iteration [Shantanu Bhadoria]
 - Initial commit [Shantanu Bhadoria]



example/Test.pl
use FindBin;
use lib "$FindBin::Bin/../lib";

use Printer::Thermal;

#Serial Printer
$printer_serial = Printer::Thermal->new( serial_device_path => '/dev/ttyS0' );

$printer_serial->test();

#usb printer
#$printer_usb = Printer::Thermal->new(usb_device_path => '/dev/usb/lp0');

#$printer_usb->test();

$printer_usb_sinocan =
  Printer::Thermal->new( usb_device_path => '/dev/ttyACM0' );

$printer_usb_sinocan->test();

#ethernet printer
$printer_ethernet =
  Printer::Thermal->new( device_ip => '192.168.168.80', device_port => '9100' );

$printer_ethernet->test();



examples/Test.pl
use strict;
use warnings;

use FindBin;
use lib "$FindBin::Bin/../lib";

use Printer::Thermal;

#Serial Printer
$printer_serial = Printer::Thermal->new( serial_device_path => '/dev/ttyS0' );

$printer_serial->test();

#usb printer
#$printer_usb = Printer::Thermal->new(usb_device_path => '/dev/usb/lp0');

#$printer_usb->test();

$printer_usb_sinocan =
  Printer::Thermal->new( usb_device_path => '/dev/ttyACM0' );

$printer_usb_sinocan->test();

#ethernet printer
$printer_ethernet =
  Printer::Thermal->new( device_ip => '192.168.168.80', device_port => '9100' );

$printer_ethernet->test();



Makefile.PL
# This Makefile.PL for  was generated by Dist::Zilla::Plugin::MakeMaker::Awesome 0.15.
# Don't edit it but the dist.ini used to construct it.

use strict;
use warnings;

use 5.010;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Interface for Thermal and some dot-matrix and inkjet Printers that support ESC-POS specification.  ",
  "AUTHOR" => "Shantanu Bhadoria (shantanu\@cpan.org)",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "Printer-Thermal",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Printer::Thermal",
  "PREREQ_PM" => {
    "Device::SerialPort" => 0,
    "IO::File" => 0,
    "IO::Socket" => 0,
    "Moose" => 0,
    "POSIX" => 0,
    "Time::HiRes" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec::Functions" => 0,
    "List::Util" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.27",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  my $tr = delete $WriteMakefileArgs{TEST_REQUIRES};
  my $br = $WriteMakefileArgs{BUILD_REQUIRES};
  for my $mod ( keys %$tr ) {
    if ( exists $br->{$mod} ) {
      $br->{$mod} = $tr->{$mod} if $tr->{$mod} > $br->{$mod};
    }
    else {
      $br->{$mod} = $tr->{$mod};
    }
  }
}

unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



